File: src/claude_memory_kit/consolidation/digest.py
import logging
from collections import defaultdict

from ..extract import consolidate_entries
from ..types import Gate

log = logging.getLogger("cmk")


async def consolidate_journals(
    db, api_key: str, user_id: str = "local"
) -> str | None:
    """Consolidate old journal entries into weekly digests."""
    stale_dates = db.stale_journal_dates(max_age_days=14, user_id=user_id)
    if not stale_dates:
        return None

    # Group by ISO week
    from datetime import datetime
    week_groups: dict[str, list[str]] = defaultdict(list)
    for date_str in stale_dates:
        dt = datetime.strptime(date_str, "%Y-%m-%d")
        week_key = dt.strftime("%G-W%V")
        week_groups[week_key].append(date_str)

    digests_written = []
    for week_key, dates in week_groups.items():
        combined = []
        for date in dates:
            entries = db.journal_by_date(date, user_id=user_id)
            for e in entries:
                combined.append(
                    f"[{e['gate']}] {e['content']}"
                )
        if not combined:
            continue

        digest = await consolidate_entries("\n".join(combined), api_key)

        # Store digest as a special journal entry
        db.insert_journal_raw(
            date=week_key,
            gate=Gate.digest,
            content=f"# Week {week_key}\n\n{digest}",
            user_id=user_id,
        )

        # Archive original entries
        for date in dates:
            db.archive_journal_date(date, user_id=user_id)

        digests_written.append(week_key)

    if not digests_written:
        return None
    return f"Consolidated {len(digests_written)} weeks: {', '.join(digests_written)}"

File: src/consolidation/digest.rs
use std::path::Path;
use anyhow::Result;
use chrono::NaiveDate;

use crate::consolidation::journal;
use crate::extract;

/// Consolidate old journal entries into a weekly digest
pub async fn consolidate_journals(
    store_path: &Path,
    api_key: &str,
) -> Result<Option<String>> {
    let stale = journal::stale_journals(store_path, 14)?;
    if stale.is_empty() {
        return Ok(None);
    }

    // Group stale journals by ISO week
    let mut week_groups: std::collections::BTreeMap<String, Vec<NaiveDate>> =
        std::collections::BTreeMap::new();
    for date in &stale {
        let week = date.format("%G-W%V").to_string();
        week_groups.entry(week).or_default().push(*date);
    }

    let mut digests_written = Vec::new();

    for (week_key, dates) in &week_groups {
        // Read all journal entries for this week
        let mut combined = String::new();
        for date in dates {
            let content = crate::store::markdown::read_journal(store_path, date)?;
            combined.push_str(&content);
            combined.push('\n');
        }

        if combined.trim().is_empty() {
            continue;
        }

        // Call Haiku to consolidate
        let digest = extract::consolidate_entries(&combined, api_key).await?;

        // Write digest file
        let digest_dir = store_path.join("digests");
        std::fs::create_dir_all(&digest_dir)?;
        let file = digest_dir.join(format!("{}.md", week_key));
        std::fs::write(&file, format!("# Week {}\n\n{}\n", week_key, digest))?;

        // Archive the original journals
        for date in dates {
            journal::archive_journal(store_path, date)?;
        }

        digests_written.push(week_key.clone());
    }

    if digests_written.is_empty() {
        Ok(None)
    } else {
        Ok(Some(format!(
            "Consolidated {} weeks: {}",
            digests_written.len(),
            digests_written.join(", ")
        )))
    }
}
